### ISSUE TITLE
0.20.13 broke Firefox/Waterfox 56 and older

### ISSUE THREAD

**Issue by mvrozanti** [akhodakivskiy/VimFx#917]
I think one of the last commits killed follow functionality.

Waterfox version 56.1.0 (64-bit)
Is it fixable?

**Comment by lydell** [comment]
I don't understand?

**Comment by rlueckl** [comment]
Same with Firefox 56.0.2 (yes, I still use it, not only because of VimFx). `f` (follow) doesn't work anymore and blocks all other shortcuts until I scroll with the mouse (`j` and `k` for example are not working after pressing `f`).

**Comment by mvrozanti** [comment]
@8 thanks! I was looking for that link. Love both VimFx and Waterfox, keep up the good work yall

**Comment by lydell** [comment]
I think this is a duplicate of #902.

**Comment by braham-snyder** [comment]
I'm not sure that it is a duplicate: while `follow` commands do not show hints anywhere after this bug occurs (`follow` commands just give `TypeError: utils.isXULDocument is not a function commands-frame.js:137:7` in the browser console), I can still scroll the page, re-order and navigate tabs through VimFx (as long as the tab I am on was created after changing the VimFx version), etc.. And removing VimFx, re-adding it, and restarting Waterfox has not worked for me.   

I am on Waterfox 56.1.0, on OS X El Capitan, with legacy extensions enabled.

<br />

On those new tabs, my own VimFx commands (`config.js` commands -- e.g. for promoting/demoting tree tabs, showing/hiding the tab bar, go-to parent) also still work.

Precise reinstall/restarting steps I tried: I downgraded to 20.12, which then worked (that is: `follow`ing worked again) only after a Waterfox restart. Then I removed VimFx, re-added 20.13, and it was broken in the same way as when I had first upgraded to 20.13. Restarting Waterfox again from there did not work either. I then removed and re-added VimFx 20.13 once more, restarted Waterfox once more, and the bug still remained. Then I removed VimFx, restarted Waterfox, and re-added VimFx 20.13, which still did not work. (Finally, I removed and re-added VimFx 20.12, which worked.)

**Comment by GloverDonovan** [comment]
8c477c0e33c2472f4fd8a13ec271fe276356f6ae seems to be the commit this regression was introduced in.

I believe that the change from `instanceof XULDocument` to `utils.isXULDocument` is the culprit here.

**Comment by iforapsy** [comment]
I had the same problem with Firefox ESR 52.7.4 (64-bit) when VimFx was automatically upgraded to     0.20.13. After the upgrade, pressing "f" to follow links did nothing visible. Downgrading to 0.20.12 fixed the problem.

**Comment by lilydjwg** [comment]
I have the same error with the latest version I've built. (The released
version does not install.) I have no idea about the error. These errors are
strange and I don't have a good way to debug. I've been trying alternatives
recently.

Simon Lydell <notifications@github.com> 于 2018年5月8日周二 04:24写道：

**Comment by jiixyj** [comment]
FWIW: Firefox fixed their usages of `instanceof XULDocument` like this:
```

     // Filter out any global which contains a XUL document.
-    if (global.document instanceof Ci.nsIDOMXULDocument) {
+    if (ChromeUtils.getClassName(global.document) == "XULDocument") {
       return false;
     }
```
It shouldn't be too hard to find a solution that works for both Firefox nightly and released versions at the same time. I can have a look later.

**Comment by lilydjwg** [comment]
I guess I've found the culprit finally.

**Comment by eli-schwartz** [comment]
I also have the issue that "/" or CTRL+F causes VimFX to enter Find mode whereby somehow it steals all keyboard input whatsoever -- I can type into the searchbox, but not the URL bar, I cannot scroll with PgUp/PgDn, I cannot exit Find mode by clicking the VimFX icon or by X-ing out the Find toolbar, and the entire tab is completely unrecoverable until I close the tab and reopen the tab from History.

I also cannot get the VimFX overlay for editing shortcuts, so I'm stuck with my current setup, but I cannot remember if that was supposed to work in the past...

I'm using firefox-developer-edition 61.0b2-1 on Arch Linux.

Is this related to the same cause and fixed by #919?

**Comment by lilydjwg** [comment]
For the find mode issue, there is #912 which should have been fixed.

You can try to build and install the fixed version if you want (setting `xpinstall.signatures.required` to false to install your own build).

**Comment by eli-schwartz** [comment]
That commit should already be in v0.20.13 which I got from AMO, so I assumed this would be a new/unfixed bug.

**Comment by lydell** [comment]
Reopening until released.

**Comment by eli-schwartz** [comment]
with #919 merged, **and** jiixyj@3b07c21726338712eb7ca964b8aa1d52995eeffd carefully cherry-picked, VimFX functions again.

**Comment by jiixyj** [comment]
Essentially, PR #920 uses the same approach of my `fix-find-mode` branch, but properly shims out getFindBar() so that older Firefox versions should continue to work.
So PR #920 instead of jiixyj@3b07c21 can be used.

**Comment by lydell** [comment]
Current status: https://github.com/akhodakivskiy/VimFx/pull/920#discussion_r187106797

Not sure what the next move is.

**Comment by lydell** [comment]
Fixed in 0.20.14.
